File: solutions/haskell/yacht/src/Yacht.hs
module Yacht (yacht, Category(..)) where

import Data.List as List

data Category = Ones
              | Twos
              | Threes
              | Fours
              | Fives
              | Sixes
              | FullHouse
              | FourOfAKind
              | LittleStraight
              | BigStraight
              | Choice
              | Yacht
    deriving Eq

diceFaceOccurences :: [Int] -> Int -> Int
diceFaceOccurences ds f = length $ filter (f ==) ds

yacht :: Category -> [Int] -> Int
yacht category dice
    | category == Ones           = diceFaceOccurences dice 1
    | category == Twos           = diceFaceOccurences dice 2 * 2
    | category == Threes         = diceFaceOccurences dice 3 * 3
    | category == Fours          = diceFaceOccurences dice 4 * 4
    | category == Fives          = diceFaceOccurences dice 5 * 5
    | category == Sixes          = diceFaceOccurences dice 6 * 6
    | category == LittleStraight && diceSorted == [1, 2, 3, 4, 5] = 30
    | category == BigStraight    && diceSorted == [2, 3, 4, 5, 6] = 30
    | category == Choice         = sum dice
    | category == Yacht          && length (nub dice) == 1 = 50
    | category == FourOfAKind =
        if      diceFaceOccurences dice 1 >= 4 then 4
        else if diceFaceOccurences dice 2 >= 4 then 8
        else if diceFaceOccurences dice 3 >= 4 then 12
        else if diceFaceOccurences dice 4 >= 4 then 16
        else if diceFaceOccurences dice 5 >= 4 then 20
        else if diceFaceOccurences dice 6 >= 4 then 24
        else 0
    | category == FullHouse = if length (nub dice) == 2 && (yacht FourOfAKind dice == 0) then sum dice else 0
    | otherwise             = 0
    where
        diceSorted = sort dice

File: solutions/python/yacht/yacht.py
YACHT = object()
ONES = object()
TWOS = object()
THREES = object()
FOURS = object()
FIVES = object()
SIXES = object()
FULL_HOUSE = object()
FOUR_OF_A_KIND = object()
LITTLE_STRAIGHT = object()
BIG_STRAIGHT = object()
CHOICE = object()

def score(dice, category):

    if category == ONES:
        return sum(1 for n in dice if n == 1)
    elif category == TWOS:
        return sum(2 for n in dice if n == 2)
    elif category == THREES:
        return sum(3 for n in dice if n == 3)
    elif category == FOURS:
        return sum(4 for n in dice if n == 4)
    elif category == FIVES:
        return sum(5 for n in dice if n == 5)
    elif category == SIXES:
        return sum(6 for n in dice if n == 6)
    elif category == LITTLE_STRAIGHT and set(dice) == {1, 2, 3, 4, 5}:
        return 30
    elif category == BIG_STRAIGHT and set(dice) == {2, 3, 4, 5, 6}:
        return 30
    elif category == CHOICE:
        return sum(dice)
    elif category == YACHT and len(set(dice)) == 1:
        return 50
    elif category == FOUR_OF_A_KIND:
        counts = { n: dice.count(n) for n in range(1, 7) }
        for face, cnt in counts.items():
            if cnt >= 4:
                return face * 4
        return 0
    elif category == FULL_HOUSE:
        dice_sorted = sorted(dice)
        a, b, c, d, e = dice_sorted

        is_three_two = (
            (a == b == c and d == e) or
            (a == b and c == d == e)
        )

        if is_three_two and len(set(dice_sorted)) == 2:
            return sum(dice_sorted)
        else:
            return 0
    else:
        return 0
